provider/core/src/main/java/org/arquillian/pact/provider/core/InteractionRunner.java

```diff
@@ -201,23 +217,30 @@ private Field validateAndGetResourceField(TestClass testClass, Class<?> fieldTyp
     }
 
     protected void executeStateChanges(final RequestResponseInteraction interaction, final TestClass testClass, final Object target) {
-        if (interaction.getProviderState() != null && !interaction.getProviderState().isEmpty()) {
-            final String state = interaction.getProviderState();
-            for (Method ann : testClass.getMethods(State.class)) {
-                if (ArrayUtils.contains(ann.getAnnotation(State.class).value(), state)) {
-                    try {
-                        ann.invoke(target);
-                    } catch (IllegalAccessException e) {
-                        throw new IllegalArgumentException(e);
-                    } catch (InvocationTargetException e) {
-                        throw new IllegalArgumentException(e);
-                    }
-                }
+        if (!interaction.getProviderStates().isEmpty()) {
+            for (final ProviderState state : interaction.getProviderStates()) {
+                Arrays.stream(testClass.getMethods(State.class))
```

Comment:
for me the thing is also about performance, doing stream for filtering something, mapping and then do the foreach or collector it might have sense, do stream.foreach and that's all as it would be the case, we are paying a performance price for not using any special stream operation.